hetmech/degree_weight.py

```diff
@@ -1,187 +1,642 @@
 import collections
+import copy
 import functools
 import itertools
+import logging
 import operator
+import time
 
 import numpy
 from scipy import sparse
 
-from .matrix import (normalize,
-                     metaedge_to_adjacency_matrix,
-                     auto_convert,
-                     copy_array)
+from .matrix import auto_convert, copy_array, metaedge_to_adjacency_matrix, \
+    normalize
 
 
-def dwwc_step(matrix, row_damping=0, column_damping=0, copy=True):
+def dwpc(graph, metapath, damping=0.5, sparse_threshold=0, use_general=False):
     """
-    Return the degree-weighted adjacency matrix produced by the input matrix
-    with the specified row and column normalization exponents.
+    A unified function to compute the degree-weighted path count.
+    This function will call get_segments, then the appropriate
+    specialized (or generalized) DWPC function.
 
     Parameters
-    ==========
-    matrix : numpy.ndarray or scipy.sparse
-        adjacency matrix for a given metaedge, where the source nodes are
-        rows and the target nodes are columns
-    row_damping : int or float
-        exponent to use in scaling each node's row by its in-degree
-    column_damping : int or float
-        exponent to use in scaling each node's column by its column-sum
-    copy : bool
-        `True` guarantees matrix will not be modified in place. `False`
-        modifies in-place if and only if matrix.dtype == numpy.float64.
-        Users are recommended not to rely on in-place conversion, but instead
-        use `False` when in-place modification is acceptable and efficiency
-        is desired.
+    ----------
+    graph : hetio.hetnet.Graph
+    metapath : hetio.hetnet.MetaPath
+    damping : float
+    sparse_threshold : float (0 <= sparse_threshold <= 1)
+        sets the density threshold above which a sparse matrix will be
+        converted to a dense automatically.
+    use_general : bool
+        if True, dwpc will call _dwpc_general_case and give a warning
+        on metapaths which are categorized 'other' and 'long_repeat'.
+        If False, an exception is raised when such a metapath is given.
 
     Returns
-    =======
-    numpy.ndarray or scipy.sparse
-        Normalized matrix with dtype.float64.
+    -------
+    numpy.ndarray, numpy.ndarray, numpy.ndarray or
+    scipy.sparse.csc_matrix, float
+        row labels, column labels, the DWPC matrix, and the computation time
     """
-    # returns a newly allocated array if copy is True
-    matrix = copy_array(matrix, copy)
+    start_time = time.perf_counter()
+    category_to_function = {'no_repeats': _dwpc_no_repeats,
```

Comment:
Tried it, and it gives errors because the functions have to all be above the dictionary definition. Since it is small, I think it may be best to leave it inside here.

Reply:
Actually, since some of the previous sub-functions need it, I may just make it into its own function.